File: IMB4 clients/Python/testclient.py
import os
import imb4

STREAM_INPUT_FILENAME = 'test.jpg'  # todo: use file name of existing file
STREAM_OUTPUT_FILENAME = 'out.python.dmp'

# define handlers for a string event, the creation of a stream and the end of a received stream
def handle_string_event(event_entry, command):
    if command == 'string command':
        print('OK received string', event_entry.event_name, command)
    else:
        print('## received string', event_entry.event_name, command)


def handle_stream_create(event_entry, stream_name):
    if stream_name == 'a stream name':
        print('OK received stream create', event_entry.event_name, stream_name)
    else:
        print('## received stream create', event_entry.event_name, stream_name)
    return open(STREAM_OUTPUT_FILENAME, 'wb')


def handle_stream_end(event_entry, stream, stream_name, cancel):
    if stream and stream_name == 'a stream name' and not cancel:
        print('OK received stream end', event_entry.event_name, stream_name, cancel)
    else:
        print('## received stream end', event_entry.event_name, stream_name, cancel)


def handle_disconnect(_):
    print('disconnected..')


# print the current folder to more easily locate stream input/output files
print('current folder', os.getcwd())

connection = imb4.TConnection(imb4.DEFAULT_REMOTE_HOST, imb4.DEFAULT_REMOTE_PORT, 'python test client', 1)
# test_connection = imb4.TConnection('192.168.1.100', imb4.DEFAULT_REMOTE_PORT, 'python test client', 1)

print('private event name', connection.private_event_name)
print('monitor event name', connection.monitor_event_name)

connection.on_disconnect = handle_disconnect

# compact statement to subscribe to int-string event and add handler
event = connection.subscribe('test event',
                             on_string_event=handle_string_event,
                             on_stream_create=handle_stream_create,
                             on_stream_end=handle_stream_end
                             )

print('subscribed to', event.event_name)

input('waiting on imb commands; press return to send events.. ')

if connection.connected:
 
    # signal string command event
    event.signal_string_event('string command')

    # signal stream
    s = open(STREAM_INPUT_FILENAME, 'rb')
    try:
        event.signal_stream('a stream name', s)
    finally:
        s.close()

    input('waiting on imb commands; press return to quit.. ')

connection.close()

File: IMB4 clients/Csharp/TestClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using IMB;

namespace TestClient
{
    class Program
    {
        static void showHelp()
        {
            Console.WriteLine(": Options");
            Console.WriteLine("   ? for help");
            Console.WriteLine("   Q or escape to quit");
            Console.WriteLine("   S to send test events");
			Console.WriteLine("   h enable heartbeat");
			Console.WriteLine("   H disable heartbeat");
            Console.WriteLine();
        }

        static void Main(string[] args)
        {
            TConnection connection = new TTLSConnection("client-eco-district.pfx", "&8dh48klosaxu90OKH", "root-ca-imb.crt", true, "C# test model");
            try
            {
                Console.WriteLine("connected");

                Console.WriteLine("private event name: " + connection.privateEventName);
                Console.WriteLine("monitor event name: " + connection.monitorEventName);

                // connect an event handler for a disconnect
                connection.onDisconnect += (aConnection) =>
                    {
                        Console.WriteLine("disconnected..");
                    };

                // connect an event handler for an exception in processing events and commands
                connection.onException += (aConnection, aException) =>
                    {
                        Console.WriteLine("## Exception: " + aException.Message);
                    };

                // subscribe to an event
                TEventEntry eventEntry = connection.subscribe("test event");

                // add an event handler for string events
                eventEntry.onString += (aEventEntry, aString) =>
                    {
                        if (aString.CompareTo("string command") == 0)
                            Console.WriteLine("OK received string " + aEventEntry.eventName + " " + aString);
                        else
                            Console.WriteLine("## received string " + aEventEntry.eventName + " " + aString);
                    };

                // add an event handler for stream create events
                eventEntry.onStreamCreate += (aEventEntry, aStreamName) =>
                    {
                        if (aStreamName == "a stream name")
                            Console.WriteLine("OK received stream create " + aEventEntry.eventName + " " + aStreamName);
                        else
                            Console.WriteLine("## received stream create " + aEventEntry.eventName + " " + aStreamName);
                        return File.Create("out.cscharp.dmp");
                    };

                // add an event handler for stream end events
                eventEntry.onStreamEnd += (aEventEntry, aStreamName, aStream, aCancel) =>
                {
                    if (aStreamName == "a stream name" && !aCancel)
                        Console.WriteLine("OK received stream end " + aEventEntry.eventName + " " + aStreamName + " " + aCancel.ToString());
                    else
                        Console.WriteLine("## received stream end " + aEventEntry.eventName + " " + aStreamName + " " + aCancel.ToString());
                };

                showHelp();

                ConsoleKeyInfo key;
                do
                {
                    key = Console.ReadKey();
                    switch (key.KeyChar)
                    {
                        case 's':
                        case 'S':
                            // send a basic string event
                            eventEntry.signalString(": string command");

                            // send a stream
                            FileStream stream = File.OpenRead("test.jpg"); // todo: use path of existing file 
                            try
                            {
                                eventEntry.signalStream("a stream name", stream);
                            }
                            finally
                            {
                                stream.Close();
                            }

                            Console.WriteLine(": sent events..");
                            break;
                        case 'h':
							connection.setHeartBeat(1000); // enable heartbeat every second
							Console.WriteLine(": heartbeat enabled..");
							break;
						case 'H':
							connection.setHeartBeat(0); // disable heartbeat (<=0)
							Console.WriteLine(": heartbeat disabled..");
							break;
						case '?':
                            showHelp();
                            break;
                    }
                }
                while (key.KeyChar != 'q' && key.KeyChar != 'Q' && key.KeyChar != (char)27);
            }
            finally
            {
                connection.close();
            }
        }
    }
}
